library/commands/qsl_commands/src/main/java/org/quiltmc/qsl/commands/impl/client/ClientCommandInternals.java

```diff
@@ -0,0 +1,285 @@
+/*
+ * Copyright 2021 QuiltMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.quiltmc.qsl.commands.impl.client;
+
+import com.google.common.collect.Iterables;
+import com.mojang.brigadier.AmbiguityConsumer;
+import com.mojang.brigadier.Command;
+import com.mojang.brigadier.CommandDispatcher;
+import com.mojang.brigadier.ParseResults;
+import com.mojang.brigadier.arguments.StringArgumentType;
+import com.mojang.brigadier.builder.ArgumentBuilder;
+import com.mojang.brigadier.builder.LiteralArgumentBuilder;
+import com.mojang.brigadier.context.CommandContext;
+import com.mojang.brigadier.context.ParsedCommandNode;
+import com.mojang.brigadier.exceptions.BuiltInExceptionProvider;
+import com.mojang.brigadier.exceptions.CommandExceptionType;
+import com.mojang.brigadier.exceptions.CommandSyntaxException;
+import com.mojang.brigadier.tree.CommandNode;
+import net.fabricmc.api.EnvType;
+import net.fabricmc.api.Environment;
+import net.minecraft.client.MinecraftClient;
+import net.minecraft.command.CommandException;
+import net.minecraft.command.CommandSource;
+import net.minecraft.text.LiteralText;
+import net.minecraft.text.Text;
+import net.minecraft.text.Texts;
+import net.minecraft.text.TranslatableText;
+import org.apache.logging.log4j.Level;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.quiltmc.qsl.commands.api.client.ClientCommandManager;
+import org.quiltmc.qsl.commands.api.client.ClientCommandRegistrationCallback;
+import org.quiltmc.qsl.commands.api.client.QuiltClientCommandSource;
+import org.quiltmc.qsl.commands.mixin.HelpCommandAccessor;
+
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+import static org.quiltmc.qsl.commands.api.client.ClientCommandManager.*;
+
+@Environment(EnvType.CLIENT)
+public final class ClientCommandInternals {
+	private static final Logger LOGGER = LogManager.getLogger();
+	private static final char PREFIX = '/';
```

Comment:
Any chance to support different prefixes for client commands? I know its not the largest issue, but I think that it does have its strengths, and there have been a few people that i have interacted with and said that they would reimplement the client command system for custom prefixes.

Reply:
For this, feel free to take a look at my original implementation for Fabric API which had custom symbol prefixes (before [this commit](https://github.com/Juuxel/fabric/commit/362b2f4c097a3367b716ab1dde5fbe2ee8b87cd7), so at [this revision](https://github.com/Juuxel/fabric/tree/148d88b8f91450510528ba1c110ec2c5b465d5e3)). (I don't mind if you take the code for QSL)

Reply:
> For this, feel free to take a look at my original implementation for Fabric API which had custom symbol prefixes (before [this commit](https://github.com/Juuxel/fabric/commit/362b2f4c097a3367b716ab1dde5fbe2ee8b87cd7), so at [this revision](https://github.com/Juuxel/fabric/tree/148d88b8f91450510528ba1c110ec2c5b465d5e3)). (I don't mind if you take the code for QSL)

Did this include completion for non-'/' commands?

Reply:
> > For this, feel free to take a look at my original implementation for Fabric API which had custom symbol prefixes (before [this commit](https://github.com/Juuxel/fabric/commit/362b2f4c097a3367b716ab1dde5fbe2ee8b87cd7), so at [this revision](https://github.com/Juuxel/fabric/tree/148d88b8f91450510528ba1c110ec2c5b465d5e3)). (I don't mind if you take the code for QSL)
> 
> Did this include completion for non-'/' commands?

Checked now, it doesn't.

I'm not a fan of per-command prefixes for multiple reasons:
 - breaks user expectations of what a command _is_ - "a command starts with `/`" is what most people first learn about commands.
 - means that we have to implement command suggestions and completion for non-`/` prefixes
 - `/` will always be special-cased _anyway_, having different code run (to prevent collisions with server commands, and also its system for command suggestion will be different to that of other prefixes)
 - extra mental load on the user, to remember which commands use which prefixes.

I wrote a partial implementation (no suggestions for non-`/` commands, slightly borked help command) based off Juuxel's work in another branch: https://github.com/williambl/quilt-standard-libraries/commit/4a994cfee06bf9a740741e4dad72fb6b3264e227.

Reply:
Yeah, they complicate the code IMO with not many benefits. I imagine the main use case is for mods grouping their commands under their own prefix. That can be done much more descriptively with a common root command (`/mod_id`), so this might not be worth it to include (especially since the autocomplete might not be fully functional).

Although... I recall having the autocomplete at least partially working in my original PR:
![](https://cdn.discordapp.com/attachments/523251999899385875/751792069403738172/unknown.png)